src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -17,52 +17,120 @@
-     */
-
-    public static final String COUNT = "count";
-    /**
-     * function that returns the maximum of a sequence
-     */
-    public static final String MIN = "min";
-    /**
-     * function that returns the minimum of a sequence
-     */
-    public static final String MAX = "max";
-    /**
-     * function that returns the minimum of a sequence
-     */
-    public static final String AVG = "avg";
-    /**
-     * function that returns the sum of a sequence
-     */
-    public static final String SUM = "sum";
-
-
-    /**
-     * function that returns the keys of a Json Object
-     */
-    public static final String KEYS = "keys";
-    /**
-     * function that returns the values of a Json Object
-     */
-    public static final String VALUES = "values";
-    /**
-     * function that returns the length of an array
-     */
-    public static final String SIZE = "size";
+    private static HashMap<SparksoniqFunctionSignature, Class<? extends RuntimeIterator>> buildInFunctions;
+
+    static {
+        buildInFunctions = new HashMap<>();
+        buildInFunctions.put(new SparksoniqFunctionSignature(JSON_FILE, 1), ParseJsonFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(JSON_FILE, 2), ParseJsonFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(PARALLELIZE, 1), ParallelizeFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(COUNT, 1), CountFunctionIterator.class);
+
+        buildInFunctions.put(new SparksoniqFunctionSignature(MIN, 1), MinFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(MAX, 1), MaxFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(SUM, 1), SumFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(AVG, 1), AvgFunctionIterator.class);
+
+        buildInFunctions.put(new SparksoniqFunctionSignature(SUBSTRING, 2), SubstringFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(SUBSTRING, 3), SubstringFunctionIterator.class);
+        for(int i = 0; i <= 100; i++)
+            buildInFunctions.put(new SparksoniqFunctionSignature(CONCAT, i), ConcatFunctionIterator.class);
+
+        buildInFunctions.put(new SparksoniqFunctionSignature(STRINGJOIN, 1), StringJoinFunction.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(STRINGJOIN, 2), StringJoinFunction.class);
+
+        buildInFunctions.put(new SparksoniqFunctionSignature(SIZE, 1), ArraySizeFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(KEYS, 1), ObjectKeysFunctionIterator.class);
+        buildInFunctions.put(new SparksoniqFunctionSignature(VALUES, 1), ObjectValuesFunctionIterator.class);
+    }
+
+    public static Class<? extends RuntimeIterator> getFunctionIteratorClass(FunctionCall expression, List<RuntimeIterator> arguments) {
```

Comment:
@ghislainfourny , this is the method used in the visitor now.

Reply:
This already looks much better. I think I may have confused you on what "signature" means. I would call the class SparksoniqFunctionNameAndArity. In JSONiq, functions are looked up by name and arity, so the map is 100% correct. There can only be one function with the same name and arity.

The signature, however, is the list of sequence types (like  integer*, object+, etc), as many sequence types as the arity of the function, and this information should also be stored somewhere, either in the map, or as additional information in the function iterator classes.

This signature information can be used to insert "treat as" iterators below function iterators when you generate the runtime iterators. A treat as expression (I am not sure if you have implemented this yet) returns exactly its input, but throws an error if the type does not match. what is expected. You then get the runtime check "for free" if the treat as iterator is correctly implemented, because anything that goes to the function iterator will have been checked by the "treat as" iterator and you no longer need to do these checks in the function iterators.

If you haven't implemented treat as expressions yet, we can postpone until later.

We can handle corner cases and casts (double to decimal, etc) later as well.